cpp/src/barretenberg/proof_system/circuit_constructors/ultra_circuit_constructor.hpp

```diff
@@ -0,0 +1,539 @@
+    void create_big_add_gate_with_bit_extraction(const add_quad& in);
+    void create_big_mul_gate(const mul_quad& in);
+    void create_balanced_add_gate(const add_quad& in);
+
+    void create_mul_gate(const mul_triple& in) override;
+    void create_bool_gate(const uint32_t a) override;
+    void create_poly_gate(const poly_triple& in) override;
+    void create_ecc_add_gate(const ecc_add_gate& in);
+
+    void fix_witness(const uint32_t witness_index, const barretenberg::fr& witness_value);
+
+    // void add_recursive_proof(const std::vector<uint32_t>& proof_output_witness_indices)
+    // {
+    //     if (contains_recursive_proof) {
+    //         failure("added recursive proof when one already exists");
+    //     }
+    //     contains_recursive_proof = true;
+
+    //     for (const auto& idx : proof_output_witness_indices) {
+    //         set_public_input(idx);
+    //         recursive_proof_public_input_indices.push_back((uint32_t)(public_inputs.size() - 1));
+    //     }
+    // }
+
+    void create_new_range_constraint(const uint32_t variable_index,
+                                     const uint64_t target_range,
+                                     std::string const msg = "create_new_range_constraint");
+    void create_range_constraint(const uint32_t variable_index, const size_t num_bits, std::string const& msg)
+    {
+        if (num_bits <= DEFAULT_PLOOKUP_RANGE_BITNUM) {
+            /**
+             * N.B. if `variable_index` is not used in any arithmetic constraints, this will create an unsatisfiable
+             *      circuit!
+             *      this range constraint will increase the size of the 'sorted set' of range-constrained integers by 1.
+             *      The 'non-sorted set' of range-constrained integers is a subset of the wire indices of all arithmetic
+             *      gates. No arithemtic gate => size imbalance between sorted and non-sorted sets. Checking for this
+             *      and throwing an error would require a refactor of the Composer to catelog all 'orphan' variables not
+             *      assigned to gates.
+             **/
+            create_new_range_constraint(variable_index, 1ULL << num_bits, msg);
+        } else {
+            decompose_into_default_range(variable_index, num_bits, DEFAULT_PLOOKUP_RANGE_BITNUM, msg);
+        }
+    }
+
+    accumulator_triple create_logic_constraint(const uint32_t a,
+                                               const uint32_t b,
+                                               const size_t num_bits,
+                                               bool is_xor_gate);
+    accumulator_triple create_and_constraint(const uint32_t a, const uint32_t b, const size_t num_bits);
+    accumulator_triple create_xor_constraint(const uint32_t a, const uint32_t b, const size_t num_bits);
+
+    uint32_t put_constant_variable(const barretenberg::fr& variable);
+
+    size_t get_num_constant_gates() const override { return 0; }
+
+    // /**
```

Comment:
You explained that you commented out these functions because they are not tested. Glad you did that, let's make sure that tests of these get written. Of course we will need to add these functions back in before removing the old composer, since without these functions we will silently break the DSL module cc @Rumata888.